src/main/java/au/org/emii/ncdfgenerator/NcdfDefinitionXMLParser.java

```diff
@@ -1,276 +1,374 @@
 
+import java.util.List;
 import java.util.ArrayList;
-import java.util.HashMap;
-import java.util.Map;
-import java.util.AbstractMap.SimpleImmutableEntry;
 
+import java.sql.Connection;
+import java.io.InputStream;
+
+import org.w3c.dom.Document;
 import org.w3c.dom.Node;
 import org.w3c.dom.NamedNodeMap;
 
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.parsers.DocumentBuilder;
+
+import java.lang.Exception ;
+
+import au.org.emii.ncdfgenerator.IDimension;
+import au.org.emii.ncdfgenerator.NcdfGeneratorException;
 
-class NcdfDefinitionXMLParser
-{
-	private boolean isNodeName( Node node, String name )
-	{
-		return node.getNodeType() == Node.ELEMENT_NODE
-			&& node.getNodeName().equals( name );
-	}
 
-	private String nodeVal( Node node  )
+class Helper
+{
+	static String nodeVal( Node node )
 	{
 		Node child = node.getFirstChild();
 		if( child != null && child.getNodeType() == Node.TEXT_NODE )
 			return child.getNodeValue();
-
 		return "";
 	}
 
-
-	private Map< String, String> parseKeyVals( Node node )
+	static boolean isElementNode( Node node )
 	{
-		// have another version that does this for attributes
-		Map< String, String> m = new HashMap< String, String>();
-		for( Node child : new NodeWrapper(node) )
-			if( child.getNodeType() == Node.ELEMENT_NODE )
-				m.put( child.getNodeName(), nodeVal( child ) );
-
-		// add any xml attributes
-		NamedNodeMap attrs = node.getAttributes();
-		for( int i = 0; i < attrs.getLength(); ++i )
-		{
-			Node child = attrs.item( i) ;
-			m.put( child.getNodeName(), nodeVal( child ) );
-		}
-
-		return m;
+	    return node.getNodeType() == Node.ELEMENT_NODE;
 	}
 
-	private IDimension parseDimension( Node node)
+/*
+	static boolean isTextNode( Node node )
 	{
-		if( isNodeName( node, "dimension")) {
-			Map< String, String> m = parseKeyVals( node );
-			return new DimensionImpl( m.get( "name" ) );
-		}
-		return null;
+	    return node.getNodeType() == Node.TEXT_NODE;
 	}
 
-	// having a simple parse key-vals function, means we can do it with attributes as alternative syntax.
+*/
+}
+
 
-	private Map< String, IDimension> parseDimensions( Node node )
+class NcdfDefinitionXMLParser
+{
+	class Context
 	{
-		if( isNodeName( node, "dimensions"))
+		// required to resolve references to dimensions from the variable definitions
+		List< IDimension> dimensions;
+
+		IDimension getDimensionByName( String name )
 		{
-			Map< String, IDimension> dimensions = new HashMap< String, IDimension> () ;
-			for( Node child : new NodeWrapper(node) ) {
-				IDimension dimension = parseDimension( child );
-				if( dimension != null)
-					dimensions.put( dimension.getName(), dimension );
+			for( IDimension dimension : dimensions ) {
+				if( dimension.getName().equals( name))
+					return dimension;
 			}
-			return dimensions;
+			return null;
 		}
-		return null;
 	}
 
 
-	private IValueEncoder parseEncoder( Node node)
-		throws NcdfGeneratorException
+	class ParseDataSource
```

Comment:
Class names in java are typically nouns, so perhaps DataSourceParser would be better.